core/o11y/src/io_tracer.rs

```diff
@@ -0,0 +1,254 @@
+    DeleteAll,
+    Other,
+}
+
+/// Formatted but not-yet printed output lines.
+///
+/// Events are bundled together and only printed after the enclosing span exits.
+/// This allows to print information at the tpo that is only available later on.
+///
+/// Note: Type used as key in `AnyMap` inside span extensions.
+struct OutputBuffer(Vec<BufferedLine>);
+
+/// Formatted but not-yet printed output line.
+struct BufferedLine {
+    indent: usize,
+    output_line: String,
+}
+
+/// Information added to a span through events happening within.
+#[derive(Default)]
+struct SpanInfo {
+    key_values: Vec<String>,
+    counts: HashMap<String, u64>,
+}
+
+impl<S: Subscriber + for<'span> LookupSpan<'span>> Layer<S> for IoTraceLayer {
+    fn on_enter(&self, id: &span::Id, ctx: tracing_subscriber::layer::Context<'_, S>) {
+        let span = ctx.span(id).unwrap();
+        span.extensions_mut().replace(OutputBuffer(vec![]));
+        span.extensions_mut().replace(SpanInfo::default());
+    }
+
+    fn on_event(&self, event: &tracing::Event<'_>, ctx: tracing_subscriber::layer::Context<'_, S>) {
+        if event.metadata().target() == "io_tracer" {
+            // Events specifically added to add more info to spans in IO Tracer.
+            // Marked with `target: "io_tracer"`.
+            if let Some(span) = ctx.event_span(event) {
+                event.record(span.extensions_mut().get_mut::<SpanInfo>().unwrap());
+            }
+        } else {
+            // All other events.
+            self.record_io_event(event, ctx);
+        }
+    }
+
+    fn on_exit(&self, id: &span::Id, ctx: tracing_subscriber::layer::Context<'_, S>) {
+        let span = ctx.span(id).unwrap();
+        let name = span.name();
+        let span_line = {
+            let mut span_info = span.extensions_mut().replace(SpanInfo::default()).unwrap();
+            for (key, count) in span_info.counts.drain() {
+                span_info.key_values.push(format!("{key}={count}"));
+            }
+            format!("{name} {}", span_info.key_values.join(" "))
+        };
+
+        let OutputBuffer(mut exiting_buffer) =
+            span.extensions_mut().replace(OutputBuffer(vec![])).unwrap();
+
+        if let Some(parent) = span.parent() {
+            let mut ext = parent.extensions_mut();
+            let OutputBuffer(parent_buffer) = ext.get_mut().unwrap();
+            parent_buffer.push(BufferedLine { indent: 2, output_line: span_line });
+            parent_buffer.extend(exiting_buffer.drain(..).map(|mut line| {
+                line.indent += 2;
+                line
+            }));
+        } else {
+            let mut out = self.file.lock().unwrap();
+            writeln!(out, "{span_line}").unwrap();
+            for BufferedLine { indent, output_line } in exiting_buffer.drain(..) {
+                writeln!(out, "{:indent$}{output_line}", "").unwrap();
+            }
+        }
+    }
+}
+
+impl IoTraceLayer {
+    pub fn new(file: Mutex<File>) -> Self {
```

Comment:
This is independent of `fmt::Json`. `MakeWriter` will construct you instances of the `io::Write` trait, making the overall subscriber generic over how it writes things to… wherever, and composable with various pre-existing high quality writer implementations such as [`NonBlocking`](https://docs.rs/tracing-appender/latest/tracing_appender/non_blocking/struct.NonBlocking.html).

Alternatively you could potentially make it generic over `io::Write` for the time being, if only to simplify the implementation effort. But you would need the `MakeWriter` if you wanted to have an ability to `lock` the mutex once per event/span rather than for each `write` operation.

Reply:
See in particular [`impl MakeWriter for Mutex<W>`](https://docs.rs/tracing-subscriber/0.3.9/src/tracing_subscriber/fmt/writer.rs.html#792-801)

Reply:
Right, added a `BufWriter` here.